Add accessibility checklist for native mobile experiences (#41)
github/annotation-toolkit: checklists/mobile-checklist.md

# Mobile Checklist

> This checklist summarizes accessibility considerations specific to native mobile experiences (iOS and Android), drawing from [Web Content Accessibility Guidelines (WCAG) 2.2](https://www.w3.org/TR/WCAG22/), the [W3C Guidance on Applying WCAG 2.2 to Mobile Applications](https://www.w3.org/TR/wcag2mobile-22/), and best practices identified by GitHub.

Reviewer: ```suggestion
This checklist summarizes accessibility considerations specific to iOS and Android, drawing from [Web Content Accessibility Guidelines (WCAG) 2.2](https://www.w3.org/TR/WCAG22/), the [W3C Guidance on Applying WCAG 2.2 to Mobile Applications](https://www.w3.org/TR/wcag2mobile-22/), and best practices identified by GitHub.
```

Death to parentheticals.

> Use this checklist when designing or building **native mobile experiences** (iOS and Android). Native mobile has its own set of accessibility patterns, APIs, and user expectations that don't map cleanly to the web, so this checklist stands on its own. You don't need to also run through the [Designer Checklist](./designer-checklist.md) or [Engineering Checklist](./engineering-checklist.md) for a native app.

Reviewer: ```suggestion
Native mobile has its own set of accessibility patterns, APIs, and user expectations that don't map cleanly to the web. Use this checklist when designing or building **native mobile experiences** on iOS and Android. You don't need to also run through the [Designer Checklist](./designer-checklist.md) or [Engineering Checklist](./engineering-checklist.md) for a native app.
```

> The one exception: if part of your app uses a **web view** (an in-app browser, an OAuth flow, embedded docs, or any HTML-rendered content), apply the web checklists to that content. Flag those areas with a [View Context Stamp](../tutorials/mobile-annotations.md#view-context-stamps-and-details) so the handoff is explicit.

Reviewer: ```suggestion
The one exception: if part of your app uses a **web view**, apply the web checklists to that content. Flag those areas with a [View Context Stamp](../tutorials/mobile-annotations.md#view-context-stamps-and-details) so the handoff is explicit. Examples of web view content are an in-app browser, an OAuth flow, embedded docs, or any HTML-rendered content.
```

> For per-interaction annotation guidance, pair this checklist with the [User Interactions tutorial](../tutorials/user-interactions.md) and the [Mobile annotations tutorial](../tutorials/mobile-annotations.md). The mobile annotations tutorial also includes a list of [design considerations](../tutorials/mobile-annotations.md#design-considerations) worth reviewing before you start annotating. For per-SC application notes when auditing, see the [Mobile-WCAG Mapping (Internal only)](https://github.com/github/accessibility-audit-guide/blob/main/mobile/mobile-wcag-map.md).

Reviewer: This reads a little rough to me at a first pass. Can we maybe break it into two paragraphs, one for per-interaction guidance and design considerations, and one for auditing notes. 

Then for each of the two new paragraphs, maybe we have a short sentence or two explaining the why behind why you'd want to do this.

- [Material Design: Accessibility](https://m3.material.io/foundations/accessible-design/overview)

---

## 1. Color

- [ ] **Meaning is not conveyed with color alone**
	- Status, severity, and selection indicators should also use a label, icon, shape, or text treatment.
- [ ] **Color contrast is observed across both light and dark modes**
	- Functional text has a **4.5:1** contrast ratio (or higher) with its background.
	- Functional graphics (icons, status indicators, control identification) have a **3:1** contrast ratio (or higher).
	- Verify both system themes since gradients and elevation tints can shift contrast.
- [ ] **Designs respect user color settings**
	- Account for high contrast modes, Smart Invert (iOS), and color correction (Android). Don't bake essential information into images that get inverted.

### Suggested Tools

- [Check color contrast - Figma Docs](https://help.figma.com/hc/en-us/articles/360041003774-Apply-paints-with-the-color-picker#h_01JQF1T71AC72G6VDXN27B77V0)
- [Web Color Contrast Checker - WebAIM](https://webaim.org/resources/contrastchecker/)

> - [Colour Contrast Analyzer for Mac and Windows - TPGi](https://www.tpgi.com/color-contrast-checker/)

Reviewer 2: Worth a mention that the Android Accessibility Scanner can detect contrast issues? It's not fool-proof, but it's decent.

---

## 2. Hierarchy

- [ ] **Each screen has at most one Title**
	- Not every screen needs a title, but if one is needed, only one is used. Titles live in the top app bar (Android) or navigation bar (iOS) and are announced on view transition. Note: in WCAG terms, [SC 2.4.2 Page Titled](https://www.w3.org/WAI/WCAG22/Understanding/page-titled.html) applies to the application as a whole on native, but the per-screen title is still the right pattern for orientation.

> - [ ] **Headings are used sparingly and only when they help structure the screen**

Reviewer 2: If a screen doesn't have a title in the top bar, I always highly encourage that the screen have some kind of prominent heading at or near the top of content.

	- Unlike the web, mobile screens often don't need headings. Use them when content has clear sections that benefit from rotor or TalkBack heading navigation.
- [ ] **Reading order matches visual order**
	- Screen readers narrate in view-tree order. Walk the screen with the VoiceOver rotor (iOS) and TalkBack swipe gestures (Android) to confirm that headings, links, form controls, and landmarks surface in the right order. This is especially important on cards, carousels, and bottom sheets.
- [ ] **Grouping is annotated**
	- Use a Mobile Grouping Stamp to mark elements that should be announced as one unit (avatar + name + timestamp on a comment, for example). This includes any content block that should read as a single unit rather than several separate stops.

### Annotations that can help
- [Mobile Structure Stamp](../tutorials/mobile-annotations.md#mobile-structure-stamp)
- [Mobile Grouping Stamp](../tutorials/mobile-annotations.md#mobile-grouping-stamp)

### Exercises

> - Sketch out the focus order with arrows before annotating. If the line zig-zags, simplify the layout or regroup elements.

Reviewer: ```suggestion
Sketch out the focus order with arrows before annotating. If the line zig-zags, simplify the layout or regroup elements.
```

Nit, but I don't like single bullet lists. They signal to screen reader users that there's more content to come, but then there isn't.

---

## 3. Content (Label, Value, Role, Hint)

- [ ] **Every interactive element has a Label**
	- Labels are the primary way assistive tech identifies a control. Match the visible text where possible so voice control users can speak what they see (per [SC 2.5.3 Label in Name](https://www.w3.org/WAI/WCAG22/Understanding/label-in-name.html)). For elements without visible text (icon-only buttons, avatars, etc.), be explicit about which label belongs to which control.
- [ ] **Stateful controls have a Value**
	- Toggles, sliders, segmented controls, and inputs need a current value (e.g., "On", "75%", "@octocat").
- [ ] **Roles and traits are specified per platform**

> 	- iOS uses traits (button, header, adjustable, selected). Android uses roles and state descriptions. Annotate both if you ship to both.

Reviewer: Kind of a general question for mobile jargon: Should we link to iOS and Android documentation for them when named?

Author: Good idea. Going to link the first mention of each throughout the checklist. Keeps things easy to scan being noisy. Maybe we can carry the pattern to the designer and engineering checklists in a follow-up.

- [ ] **Hints are used only for non-obvious interactions**
	- Don't put critical info in hints. Users can disable them, and they're announced last.
- [ ] **Decorative elements are hidden from assistive tech**

> 	- Mark with `accessibilityElementsHidden` (iOS) or `importantForAccessibility="no"` (Android).

Reviewer: Maybe link to an explainer about what constitutes as decorative?

> 	- Mark with `accessibilityElementsHidden` (iOS) or `importantForAccessibility="no"` (Android).

Reviewer 2: Rather than `accessibilityElementsHidden`, I'd recommending citing `.accessibilityHidden` set to `true`, which is specific to iOS (SwiftUI specifically).

- [ ] **Language of the app is set, and any sections in a different language are identified**

> 	- On native, [SC 3.1.1 Language of Page](https://www.w3.org/WAI/WCAG22/Understanding/language-of-page.html) applies to the application as a whole, not individual screens. For inline text in a different language, identify it programmatically per [SC 3.1.2 Language of Parts](https://www.w3.org/WAI/WCAG22/Understanding/language-of-parts.html).

Reviewer 2: Consider adding a reference link for this, as it's kind of varsity-level. Appt.org has a good starting point that covers all major native mobile frameworks: https://appt.org/en/guidelines/wcag/success-criterion-3-1-2

- [ ] **Auto-playing or scrolling content can be paused, stopped, or hidden**
	- Includes auto-advancing carousels, looping animations, and live previews.
- [ ] **Nothing flashes more than 3 times per second**

### Resources

- [Alt Text Guide - GitHub Workplace Accessibility](https://github.com/github/workplace-accessibility/blob/main/resources/content-accessibility/alt-text-guide.md)
- [W3C alt-text decision tree](https://www.w3.org/WAI/tutorials/images/decision-tree/)

---

## 5. Interactivity and touch targets

- [ ] **Touch targets are at least 44x44 pt (iOS) / 48x48 dp (Android)**
	- This meets [WCAG 2.2 SC 2.5.8 Target Size (Minimum)](https://www.w3.org/WAI/WCAG22/Understanding/target-size-minimum.html) of 24x24 CSS pixels with healthy margin, and aligns with both platforms' HIG guidance.
	- Targets that let users select a value spatially based on position within the target (sliders, color pickers, map pins inside a draggable region) are considered a single target for this SC.
- [ ] **Adjacent targets have enough spacing**
	- Aim for at least 8 pt/dp between tappable elements to prevent mis-taps.
- [ ] **Buttons and links are used appropriately**

> 	- Buttons trigger actions. Links navigate. Don't style one as the other. When annotating, identify every element that performs an action when activated so the right role is applied.

Reviewer 2: It's less clearcut on mobile -- the general practice I teach and that most platform apps seem to follow is that anything that navigates you within an app, or that triggers a system function (e.g. initiating a phone call) is a button. Links are reserved for content that opens in the device browser, or possibly web views.

- [ ] **Each button and link has a unique, descriptive name**
	- Avoid repeated "More", "Open", or "View" labels in lists. Include context (e.g., "Open pull request 142").
- [ ] **Disabled buttons are avoided**
	- Prefer an inactive style with feedback explaining what's missing, or hide the action until it's available.

### Annotations that can help
- [Mobile Button Stamp](../tutorials/mobile-annotations.md#mobile-button-stamp)

### Resources

- [Getting To The Bottom Of Minimum WCAG-Conformant Interactive Element Size - Smashing Magazine](https://www.smashingmagazine.com/2024/07/getting-bottom-minimum-wcag-conformant-interactive-element-size/)

> - [Apple HIG: Buttons](https://developer.apple.com/design/human-interface-guidelines/buttons)

Reviewer 2: This link covers target size on a platform rather than component level: https://developer.apple.com/design/human-interface-guidelines/accessibility#Mobility

- [Material Design: Touch targets](https://m3.material.io/foundations/designing/structure#1da3138c-2b54-4f5b-9c88-dbf6e6b9c12a)

---

## 6. Gestures and motion

> The [User Interactions tutorial](../tutorials/user-interactions.md#touch-gesture) groups touch gestures into three tiers: **Basic** (single tap, double tap, long press), **Specialized** (swipe, drag, tap-and-hold), and **Advanced** (multi-finger, pinch, rotate, path-based). Use this section to verify each tier is used appropriately.

Reviewer: ```suggestion
The [User Interactions tutorial](../tutorials/user-interactions.md#touch-gesture) groups touch gestures into three tiers: 

1. **Basic** (single tap, double tap, long press), 
2. **Specialized** (swipe, drag, tap-and-hold), and 
3. **Advanced** (multi-finger, pinch, rotate, path-based). 

Use this section to verify each tier is used appropriately.
```

- [ ] **Primary actions use Basic gestures**

> 	- Single tap is the most accessible. Reserve it for the most important interactions on a screen.

Reviewer 2: "Reserve it for the most important interactions on a screen" feels off given that single tap is the foundational touch gesture and interactive elements assume single tap unless otherwise specified.

> - [ ] **Specialized gestures are used sparingly and have a single-point alternative**

Reviewer 2: "single tap" alternative -- "single point" is pretty jargony

- [ ] **Pointer cancellation is supported**
	- Per [SC 2.5.2 Pointer Cancellation](https://www.w3.org/WAI/WCAG22/Understanding/pointer-cancellation.html), trigger actions on the up-event so users can drag off a control to cancel.

### Annotations that can help
- [Touch gesture (User Interactions)](../tutorials/user-interactions.md#touch-gesture)
- [User Interactions: Custom Action](https://gh.io/annotation-tutorial-user-interactions)

---

## 7. Device settings and platform behavior

Mobile users rely heavily on system-level settings to make their device usable. The [Device setting annotations](../tutorials/user-interactions.md#device-setting) cover the most common ones. Designs need to respect these settings, not fight them.

- [ ] **Both portrait and landscape orientations are supported**
	- Per [SC 1.3.4 Orientation](https://www.w3.org/WAI/WCAG22/Understanding/orientation.html), don't lock orientation unless it's essential (e.g., a piano app). Many users mount their device in a fixed orientation for accessibility reasons.

> - [ ] **Layouts adapt to Dynamic Type (iOS) and Font size scaling (Android)**

Reviewer: ```suggestion
- [ ] **Layouts adapt to iOS' Dynamic Type and Android's Font Size Scaling**
```

	- Test at the largest accessibility text sizes per [SC 1.4.4 Resize Text](https://www.w3.org/WAI/WCAG22/Understanding/resize-text.html). Content should reflow without truncation or overlap.
- [ ] **Viewport zoom and viewport resize don't break the layout**
	- Magnifier, browser zoom inside web views, and split-screen / multitasking on tablets all change the available viewport. No content overlap, no obscured controls.
- [ ] **Reduce Motion is honored**
	- Annotate any motion-heavy transition with the Reduced motion device setting so engineering knows to gate it.
- [ ] **Voice Control labels match visible text**

> 	- Voice Control (iOS) and Voice Access (Android) let users speak the name of a control. If your label and visible text drift apart, voice users can't activate it.

Reviewer 2: For iOS only, you can also define alternate .accessibilityInputLabels that Voice Control users can speak if there may be more than one common, expected name, or if there's a conflict between a very long accessibilityLabel that a VO user needs to hear and the shorter/simpler label that a Voice Control user may need/want to say.

For instance, a trash can icon button's label could be "Delete" and it could be set to also accept "trash" and "garbage" when someone speaks those words in a command. This might be too detailed for this checklist, but adding alternate labels makes a huge difference in the quality of the Voice Control experience if your app has lots of long accessibility labels.

	- On native, [SC 1.4.10 Reflow](https://www.w3.org/WAI/WCAG22/Understanding/reflow.html) applies through WCAG2ICT guidance: when users resize text via Dynamic Type or font scaling, content should reflow without requiring scrolling in two dimensions for primary content. Tables, image galleries, and similar are reasonable exceptions.
- [ ] **Persistent UI (nav, help, status) appears in consistent locations across screens**
	- Per [SC 3.2.3 Consistent Navigation](https://www.w3.org/WAI/WCAG22/Understanding/consistent-navigation.html) and [SC 3.2.6 Consistent Help](https://www.w3.org/WAI/WCAG22/Understanding/consistent-help.html).
- [ ] **Web view content (only) supports text spacing overrides**
	- [SC 1.4.12 Text Spacing](https://www.w3.org/WAI/WCAG22/Understanding/text-spacing.html) applies only inside a web view on native. For embedded HTML content, verify line height up to 1.5x font size, paragraph spacing up to 2x, letter spacing up to 0.12x, and word spacing up to 0.16x. Native UI is governed by Dynamic Type / font scaling instead (see §7).

---

## 10. Touch and keyboard navigation

- [ ] **Every touch interaction is reachable with an external keyboard**

> 	- Bluetooth keyboards, Switch Control (iOS), and Switch Access (Android) all rely on a sensible focus order. Per [SC 2.1.1 Keyboard](https://www.w3.org/WAI/WCAG22/Understanding/keyboard.html), the keyboard interface here means any keyboard or keyboard substitute, not just a physical device.

Reviewer: ```suggestion
	- Bluetooth keyboards, iOS Switch Control, and Android Switch Access all rely on a sensible focus order. Per [SC 2.1.1 Keyboard](https://www.w3.org/WAI/WCAG22/Understanding/keyboard.html), the keyboard interface here means any keyboard or keyboard substitute, not just a physical device.
```

- [ ] **A visible focus indicator appears when navigating with a keyboard or switch**
	- Per [SC 2.4.7 Focus Visible](https://www.w3.org/WAI/WCAG22/Understanding/focus-visible.html) and [SC 2.4.11 Focus Not Obscured (Minimum)](https://www.w3.org/WAI/WCAG22/Understanding/focus-not-obscured-minimum.html).

> - [ ] **Focus order is annotated and matches the visual flow**

Reviewer 2: Perhaps position this as more of an exception -- annotate focus order when there's a complex layout that needs it. Being overly prescriptive with focus order (especially on simple screens) can be a real footgun depending on how it's executed.

	- Avoid focus that jumps back and forth across the screen.
- [ ] **There are no focus traps**
	- Modals, sheets, and overlays must let users dismiss and return focus to a logical place. Per [SC 2.1.2 No Keyboard Trap](https://www.w3.org/WAI/WCAG22/Understanding/no-keyboard-trap.html).
- [ ] **Custom controls expose proper accessibility actions**

> 	- A custom slider should respond to increment/decrement actions, not just drag.

Reviewer 2: "For example, a custom slider..."

### Annotations that can help
- [Ordering](https://gh.io/annotation-tutorial-ordering) (Focus Order, Arrow Stop)
- [Keyboard shortcut (User Interactions)](../tutorials/user-interactions.md#keyboard-shortcut)

---

## 11. Platform parity

- [ ] **Stock platform controls are preferred over custom ones**
	- A native segmented control, switch, or date picker comes with accessibility behavior built in. Before designing a custom control, check whether a standard one would do the job.
- [ ] **iOS and Android specs are both annotated when shipping to both platforms**
	- Use the iOS (blue) and Android (green) variants of mobile annotations to make platform differences visible on the canvas.
- [ ] **Stock components use the platform's native accessibility behavior**
	- When using a system component, toggle "Follow stock native pattern" in your Mobile Details annotation and link to the platform docs.
- [ ] **Platform conventions are respected**

> 	- Back navigation, modal dismissal, and primary action placement differ between iOS and Android. Don't force one platform's pattern onto the other.

Reviewer 2: Worth mentioning the VO scrub gesture here, that behaves similar to the back button? It can navigate backward, dismiss modals, and be assigned to collapse/hide custom components when it makes sense. If you're not using stock components, users will expect support of the scrub gesture and it has to be added manually.

---

## 12. Platform functions

The [Platform function annotations](../tutorials/user-interactions.md#platform-function) flag built-in behaviors that can disorient users if they're not designed carefully.

- [ ] **Loading states have clear visual and programmatic feedback**
	- Use a live region announcement, not just a spinner. See [Mobile Live Region Announcement](../tutorials/mobile-annotations.md#mobile-live-region-announcement). Per [SC 4.1.3 Status Messages](https://www.w3.org/WAI/WCAG22/Understanding/status-messages.html).
- [ ] **Automatic transitions are avoided or controllable**
	- System-initiated screen, state, or content changes should be paused, slowed, or dismissable. Common offenders: onboarding carousels, auto-advancing tutorials. Per [SC 3.2.1 On Focus](https://www.w3.org/WAI/WCAG22/Understanding/on-focus.html) and [SC 3.2.2 On Input](https://www.w3.org/WAI/WCAG22/Understanding/on-input.html).
- [ ] **Opening external content is announced**

> 	- "Opens in Safari", "Opens in browser", or "Opens in a new window" warnings prevent disorientation, especially when leaving an app for a web view.

Reviewer 2: Recommend striking the Safari example since iPhone users can customize their default browser (even if it is all just Safari/Webkit  under the hood 🙃 )

- [ ] **Time limits can be turned off, adjusted, or extended**
	- Exceptions: real-time events (auctions, live games) or time limits beyond 20 hours. Per [SC 2.2.1 Timing Adjustable](https://www.w3.org/WAI/WCAG22/Understanding/timing-adjustable.html).
- [ ] **Cognitive function tests have an accessible alternative**
	- CAPTCHA puzzles, math challenges, and memory checks need an alternate path. Per [SC 3.3.8 Accessible Authentication (Minimum)](https://www.w3.org/WAI/WCAG22/Understanding/accessible-authentication-minimum.html). Note: passwords used to unlock the underlying platform (device passcode, biometric unlock) are out of scope for this requirement at the app level.

### Annotations that can help
- [Platform function (User Interactions)](../tutorials/user-interactions.md#platform-function): Loading, Automatic transition, Open in new tab, Time limit, Cognitive puzzle
- [Mobile Live Region Announcement](../tutorials/mobile-annotations.md#mobile-live-region-announcement)

---

## 13. Notifications and live updates

> - [ ] **Status changes are announced via Live Regions**

Reviewer 2: ~Live Regions~ live accessibility announcements
(live regions aren't really a thing, although we can accomplish similar goals. Just different terms.)

### Suggested Tools

- [Check color contrast - Figma Docs](https://help.figma.com/hc/en-us/articles/360041003774-Apply-paints-with-the-color-picker#h_01JQF1T71AC72G6VDXN27B77V0)
- [Web Color Contrast Checker - WebAIM](https://webaim.org/resources/contrastchecker/)
- [Colour Contrast Analyzer for Mac and Windows - TPGi](https://www.tpgi.com/color-contrast-checker/)

---

## 2. Hierarchy

- [ ] **Each screen has at most one Title**
	- Not every screen needs a title, but if one is needed, only one is used. Titles live in the top app bar (Android) or navigation bar (iOS) and are announced on view transition. Note: in WCAG terms, [SC 2.4.2 Page Titled](https://www.w3.org/WAI/WCAG22/Understanding/page-titled.html) applies to the application as a whole on native, but the per-screen title is still the right pattern for orientation.
- [ ] **Headings are used sparingly and only when they help structure the screen**
	- Unlike the web, mobile screens often don't need headings. Use them when content has clear sections that benefit from rotor or TalkBack heading navigation.
- [ ] **Reading order matches visual order**

> 	- Screen readers narrate in view-tree order. Walk the screen with the VoiceOver rotor (iOS) and TalkBack swipe gestures (Android) to confirm that headings, links, form controls, and landmarks surface in the right order. This is especially important on cards, carousels, and bottom sheets.

Reviewer 3: I’ve completed some work with React Native developers recently, and we spent some time understanding how the reading order on mobile works. As explained in the opening sentence on this line, there’s no way for React Native developers to change the reading order - the operating system scans the content on screen and then provides it to the screen readers *roughly* from top to bottom, left to right. This can sometimes lead to slightly unexpected results, depending on the placement of content.

Therefore, I wondered if it’s worth having a point focused on the structure of the design -something that raises this as a point to consider before people even get to testing so that this can be incorporated in to any designs, particularly given that this document specifies that people reading it don’t need to consult the design checklist?

In terms of what this might be - maybe something to say that depending on the framework used for development, it may not be possible to change the reading order, so try to utilise simple, easy to follow layouts to mitigate the risk of screen readers not announcing content in the expected order?

- [ ] **Grouping is annotated**
	- Use a Mobile Grouping Stamp to mark elements that should be announced as one unit (avatar + name + timestamp on a comment, for example). This includes any content block that should read as a single unit rather than several separate stops.

### Annotations that can help
- [Mobile Structure Stamp](../tutorials/mobile-annotations.md#mobile-structure-stamp)
- [Mobile Grouping Stamp](../tutorials/mobile-annotations.md#mobile-grouping-stamp)

### Exercises

Sketch out the focus order with arrows before annotating. If the line zig-zags, simplify the layout or regroup elements.

---

## 3. Content (Label, Value, Role, Hint)

- [ ] **Every interactive element has a Label**

> 	- Labels are the primary way assistive tech identifies a control. Match the visible text where possible so voice control users can speak what they see (per [SC 2.5.3 Label in Name](https://www.w3.org/WAI/WCAG22/Understanding/label-in-name.html)). For elements without visible text (icon-only buttons, avatars, etc.), be explicit about which label belongs to which control.

Reviewer 3: This is great, there are a few different points in here, so I wondered if it might be helpful to break it down in to the separate concerns, maybe something along the lines of:

> - Labels help people understand the purpose of a control, and while useful for everyone, they’re particularly important for people with thinking disabilities.
> - For people using voice control, ensure that the visible label text is programmatically related to the control, so that people can speak the name of the control that they see (per [SC 2.5.3 Label in Name](https://www.w3.org/WAI/WCAG22/Understanding/label-in-name.html).
> - For elements without visible text (icon-only buttons, avatars, etc.), be explicit about which label belongs to which control.

Please feel free to take/dismiss as seems useful 😅

Author: Changed to

> - [ ] **Every interactive element has a Label**
	- Labels help people understand the purpose of a control. They're useful for everyone and especially important for people with cognitive disabilities.
	- For Voice Control and Voice Access users, the visible label text must be programmatically associated with the control so users can speak the name they see. Per [SC 2.5.3 Label in Name](https://www.w3.org/WAI/WCAG22/Understanding/label-in-name.html).
	- For elements without visible text (icon-only buttons, avatars, etc.), be explicit in your annotations about which label belongs to which control.

- [ ] **Stateful controls have a Value**
	- Toggles, sliders, segmented controls, and inputs need a current value (e.g., "On", "75%", "@octocat").
- [ ] **Roles and traits are specified per platform**
	- iOS uses traits (button, header, adjustable, selected). Android uses roles and state descriptions. Annotate both if you ship to both.
- [ ] **Hints are used only for non-obvious interactions**

> 	- Don't put critical info in hints. Users can disable them, and they're announced last.

Reviewer 3: This is an excellent point to raise as its something that feels like its missed quite often, so great to include it here!!

- [ ] **Auto-playing or scrolling content can be paused, stopped, or hidden**
	- Includes auto-advancing carousels, looping animations, and live previews.
- [ ] **Nothing flashes more than 3 times per second**

### Resources

- [Alt Text Guide - GitHub Workplace Accessibility](https://github.com/github/workplace-accessibility/blob/main/resources/content-accessibility/alt-text-guide.md)
- [W3C alt-text decision tree](https://www.w3.org/WAI/tutorials/images/decision-tree/)

---

## 5. Interactivity and touch targets

- [ ] **Touch targets are at least 44x44 pt (iOS) / 48x48 dp (Android)**
	- This meets [WCAG 2.2 SC 2.5.8 Target Size (Minimum)](https://www.w3.org/WAI/WCAG22/Understanding/target-size-minimum.html) of 24x24 CSS pixels with healthy margin, and aligns with both platforms' HIG guidance.
	- Targets that let users select a value spatially based on position within the target (sliders, color pickers, map pins inside a draggable region) are considered a single target for this SC.
- [ ] **Adjacent targets have enough spacing**
	- Aim for at least 8 pt/dp between tappable elements to prevent mis-taps.
- [ ] **Buttons and links are used appropriately**

> 	- Buttons trigger actions. Links navigate. Don't style one as the other. When annotating, identify every element that performs an action when activated so the right role is applied.

Reviewer 3: This is really important! I wondered if the last sentence could be adjusted ever so slightly to make the consideration around the importance of choosing the correct role, a little more clear. To help illustrate what I mean, hopefully this example helps:

> When annotating, consider the action performed when interacting with each element, and ensure that the right role is applied.

- [ ] **Pointer cancellation is supported**
	- Per [SC 2.5.2 Pointer Cancellation](https://www.w3.org/WAI/WCAG22/Understanding/pointer-cancellation.html), trigger actions on the up-event so users can drag off a control to cancel.

### Annotations that can help
- [Touch gesture (User Interactions)](../tutorials/user-interactions.md#touch-gesture)
- [User Interactions tutorial](https://gh.io/annotation-tutorial-user-interactions)

---

## 7. Device settings and platform behavior

Mobile users rely heavily on system-level settings to make their device usable. The [Device setting annotations](../tutorials/user-interactions.md#device-setting) cover the most common ones. Designs need to respect these settings, not fight them.

- [ ] **Both portrait and landscape orientations are supported**
	- Per [SC 1.3.4 Orientation](https://www.w3.org/WAI/WCAG22/Understanding/orientation.html), don't lock orientation unless it's essential (e.g., a piano app). Many users mount their device in a fixed orientation for accessibility reasons.
- [ ] **Layouts adapt to iOS' Dynamic Type and Android's Font Size Scaling**

> 	- Test at the largest accessibility text sizes per [SC 1.4.4 Resize Text](https://www.w3.org/WAI/WCAG22/Understanding/resize-text.html). Content should reflow without truncation or overlap.

Reviewer 3: I wonder if it would also be helpful to specify that all functionality and content must remain available and must not be lost?

- [ ] **Viewport zoom and viewport resize don't break the layout**

> 	- Magnifier, browser zoom inside web views, and split-screen / multitasking on tablets all change the available viewport. No content overlap, no obscured controls.

Reviewer 3: I’ve never seen anyone make the point around multitasking on tablets in relation to accessibility before! This is a great shout, I’ll be using this example from now on 😄

- [ ] **Reduce Motion is honored**
	- Annotate any motion-heavy transition with the Reduced motion device setting so engineering knows to gate it.
- [ ] **Voice Control labels match visible text**

> 	- Voice Control (iOS) and Voice Access (Android) let users speak the name of a control. If your label and visible text drift apart, voice users can't activate it.

Reviewer 3: I wonder if drift apart could be unpacked to help make the visual and programmatic distinction a little more clear? Maybe something along the lines of:

> If your label and visible text aren’t visually related and programmatically related, voice users can't activate it.

Hopefully this will help to highlight both aspects of the consideration that must be accounted for 😅

- [ ] **Functionality does not require device motion to operate**
	- Shake-to-undo, tilt-to-scroll, and similar interactions need a button equivalent and should be possible to disable. Per [SC 2.5.4 Motion Actuation](https://www.w3.org/WAI/WCAG22/Understanding/motion-actuation.html).

### Annotations that can help
- [Device setting (User Interactions)](../tutorials/user-interactions.md#device-setting): Reduced motion, Viewport zoom, Voice input, Viewport resize, Orientation, Shake device

---

## 8. Forms

- [ ] **Inputs have a persistent visible label, not just a placeholder**
	- Per [SC 3.3.2 Labels or Instructions](https://www.w3.org/WAI/WCAG22/Understanding/labels-or-instructions.html).
- [ ] **Keyboard type matches the input**

> 	- Email, number, phone, URL keyboards reduce errors and effort.

Reviewer 3: There’s this GOV.UK blog post (which it shocked me when checking it just now to see that it’s 6 years old 🤯 , so it may not be pertinent): [Why the GOV.UK Design System team changed the input type for numbers](https://technology.blog.gov.uk/2020/02/24/why-the-gov-uk-design-system-team-changed-the-input-type-for-numbers/), which covers why the number input might be difficult to use. I appreciate that it’s difficult to explain nuances in a checklist that is trying to be as comprehensive and relatable as this, so it might be worth ignoring this!

Author: Really appreciate this — it's a great post. I think we might be able to find a spot for it in the Mobile tutorial doc rather than the checklist itself. The nuance feels too deep for this particular format.

Reviewer 3: No problem at all, that makes sense to me!!

Reviewer 2: In native mobile, the input type is just about the keyboard. Identifying a field as using the numeric keyboard has no impact on the field's behavior. To achieve the kind of spinner behaviour that can be problematic on the web, you'd have to use a stepper in iOS or build a custom stepper control in Android.

TL;DR: the Gov.UK article has good nuance to consider for web but the problems it considers don't exist in native mobile.

Reviewer 3: Ah, apologies - I think I must have conflated that article with native mobile because of the general mobile connection 😅, sorry about this and thank you for catching it @julianmka! 🙌

- [ ] **Autofill, autocomplete, and content type hints are set**
	- iOS `textContentType`, Android `autofillHints`. Helps password managers and reduces redundant entry per [SC 3.3.7 Redundant Entry](https://www.w3.org/WAI/WCAG22/Understanding/redundant-entry.html) and [SC 1.3.5 Identify Input Purpose](https://www.w3.org/WAI/WCAG22/Understanding/identify-input-purpose.html).
- [ ] **Errors are announced and tied to the field**

> 	- Per [SC 3.3.1 Error Identification](https://www.w3.org/WAI/WCAG22/Understanding/error-identification.html) and [SC 3.3.3 Error Suggestion](https://www.w3.org/WAI/WCAG22/Understanding/error-suggestion.html). Use live regions or accessibility focus to surface errors. Don't rely on color alone.

Reviewer 3: I think this should also explain that errors must be indicated using text, and that if it’s possible to determine what a user needs to do to resolve the error, this should also be provided?

- [ ] **Content reflows without loss of information or functionality**
	- On native, [SC 1.4.10 Reflow](https://www.w3.org/WAI/WCAG22/Understanding/reflow.html) applies through WCAG2ICT guidance: when users resize text via Dynamic Type or font scaling, content should reflow without requiring scrolling in two dimensions for primary content. Tables, image galleries, and similar are reasonable exceptions.
- [ ] **Persistent UI (nav, help, status) appears in consistent locations across screens**
	- Per [SC 3.2.3 Consistent Navigation](https://www.w3.org/WAI/WCAG22/Understanding/consistent-navigation.html) and [SC 3.2.6 Consistent Help](https://www.w3.org/WAI/WCAG22/Understanding/consistent-help.html).
- [ ] **Web view content (only) supports text spacing overrides**
	- [SC 1.4.12 Text Spacing](https://www.w3.org/WAI/WCAG22/Understanding/text-spacing.html) applies only inside a web view on native. For embedded HTML content, verify line height up to 1.5x font size, paragraph spacing up to 2x, letter spacing up to 0.12x, and word spacing up to 0.16x. Native UI is governed by Dynamic Type / font scaling instead (see §7).

---

## 10. Touch and keyboard navigation

> - [ ] **Every touch interaction is reachable with an external keyboard**

Reviewer 3: Super super minor, but I think it might be helpful to clarify that the interactions must be both reachable and interactive using only an external keyboard - just so that people using the checklist also know to test the interaction too!

### Annotations that can help
- [Ordering](https://gh.io/annotation-tutorial-ordering) (Focus Order, Arrow Stop)
- [Keyboard shortcut (User Interactions)](../tutorials/user-interactions.md#keyboard-shortcut)

---

## 11. Platform parity

- [ ] **Stock platform controls are preferred over custom ones**
	- A native segmented control, switch, or date picker comes with accessibility behavior built in. Before designing a custom control, check whether a standard one would do the job.
- [ ] **iOS and Android specs are both annotated when shipping to both platforms**
	- Use the iOS (blue) and Android (green) variants of mobile annotations to make platform differences visible on the canvas.
- [ ] **Stock components use the platform's native accessibility behavior**
	- When using a system component, toggle "Follow stock native pattern" in your Mobile Details annotation and link to the platform docs.
- [ ] **Platform conventions are respected**

> 	- Back navigation, modal dismissal, and primary action placement differ between iOS and Android. Don't force one platform's pattern onto the other.

Reviewer 3: This is a really important point for usability, so it's great to see it included here! 🙌

- [ ] **Color contrast is observed across both light and dark modes**
	- Functional text has a **4.5:1** contrast ratio (or higher) with its background.
	- Functional graphics (icons, status indicators, control identification) have a **3:1** contrast ratio (or higher).
	- Verify both system themes since gradients and elevation tints can shift contrast.
- [ ] **Designs respect user color settings**
	- Account for high contrast modes, [Smart Invert (iOS)](https://support.apple.com/guide/iphone/change-display-and-text-size-iph3e2e1fb0/ios), and [color correction (Android)](https://support.google.com/accessibility/android/answer/11183305?hl=en-GB&sjid=1525872318388975084-NC#zippy=%2Cuse-colour-correction:~:text=Colour%20correction%20and%20grayscale%20settings%20help%20your%20device%20compensate%20for%20colour%20blindness.%C2%A0). Don't bake essential information into images that get inverted.

### Suggested Tools

- [Check color contrast - Figma Docs](https://help.figma.com/hc/en-us/articles/360041003774-Apply-paints-with-the-color-picker#h_01JQF1T71AC72G6VDXN27B77V0)
- [Web Color Contrast Checker - WebAIM](https://webaim.org/resources/contrastchecker/)

> - [Colour Contrast Analyzer for Mac and Windows - TPGi](https://www.tpgi.com/color-contrast-checker/)

Reviewer 3: Only just caught this, but as a heads up, TPGi are now Vispero - so I think this should be:

```
- [Colour Contrast Analyzer for Mac and Windows - Vispero](https://vispero.com/lp/color-contrast-checker/)
```

Author: Good catch!

# Mobile Checklist

This checklist summarizes accessibility considerations specific to iOS and Android, drawing from [Web Content Accessibility Guidelines (WCAG) 2.2](https://www.w3.org/TR/WCAG22/), the [W3C Guidance on Applying WCAG 2.2 to Mobile Applications](https://www.w3.org/TR/wcag2mobile-22/), and best practices identified by GitHub.

> Native mobile has its own set of accessibility patterns, APIs, and user expectations that don't map cleanly to the web. Use this checklist when designing or building **native mobile experiences** on iOS and Android. You don't need to also run through the [Designer Checklist](./designer-checklist.md) or [Engineering Checklist](./engineering-checklist.md) for a native app.

Reviewer 2: Add something about how the underlying user needs to perceive, operate, and understand experiences are the same, despite the differences in the platform.

- [ ] **If a screen has no title bar, include a prominent heading near the top of the content**
	- Screens without a top app bar (Android) or navigation bar (iOS) should still orient users — a clear, prominent heading at or near the top of the visible content gives screen reader users and everyone else a place to land.
- [ ] **Headings are used sparingly and only when they help structure the screen**
	- Unlike the web, mobile screens often don't need headings. Use them when content has clear sections that benefit from rotor or [TalkBack](https://support.google.com/accessibility/android/answer/6283677) heading navigation.
- [ ] **Layouts favor a simple, predictable flow**
	- On some frameworks like React Native, the operating system determines reading order by scanning the screen roughly top-to-bottom, left-to-right, and there's no API to override it. Designs that follow a clear vertical flow narrate correctly regardless of framework, and they're easier for everyone to scan visually too.
- [ ] **Reading order matches visual order**

> 	- Screen readers narrate in view-tree order. Walk the screen with the [VoiceOver](https://support.apple.com/guide/iphone/turn-on-and-practice-voiceover-iph3e2e415f/ios) rotor (iOS) and TalkBack swipe gestures (Android) to confirm that headings, links, form controls, and landmarks surface in the right order. This is especially important on cards, carousels, and bottom sheets.

Reviewer 2: Missed this earlier, but take out landmarks. iOS has containers, technically, but they're pretty useless for navigating within an app.

	- For elements without visible text (icon-only buttons, avatars, etc.), be explicit in your annotations about which label belongs to which control.
- [ ] **Stateful controls have a Value**
	- Toggles, sliders, segmented controls, and inputs need a current value (e.g., "On", "75%", "@octocat").
- [ ] **Roles and traits are specified per platform**
	- iOS uses traits (button, header, adjustable, selected). Android uses roles and state descriptions. Annotate both if you ship to both.
- [ ] **Hints are used only for non-obvious interactions**
	- Don't put critical info in hints. Users can disable them, and they're announced last.
- [ ] **Directionality is not used in content**
	- Avoid wording like "see below" or "to the left/right" — content positions shift across orientations, screen sizes, and reading order. Use "first/last" or "previous/next" instead.
- [ ] **Truncation is only used when unavoidable**

> 	- Truncated text hides information and can break for users at larger Dynamic Type sizes. Prefer reflowing, wrapping, or progressive disclosure (expand/collapse) over truncation. When truncation is unavoidable, ensure the full content is reachable some other way.

Reviewer 2: "Larger font sizes" - Dynamic Type only applies to iOS

- [ ] **Language of the app is set, and any sections in a different language are identified**
	- On native, [SC 3.1.1 Language of Page](https://www.w3.org/WAI/WCAG22/Understanding/language-of-page.html) applies to the application as a whole, not individual screens. For inline text in a different language, identify it programmatically per [SC 3.1.2 Language of Parts](https://www.w3.org/WAI/WCAG22/Understanding/language-of-parts.html). For implementation details across native frameworks, see [Appt's SC 3.1.2 guidance](https://appt.org/en/guidelines/wcag/success-criterion-3-1-2).

### Annotations that can help
- [Mobile Details](../tutorials/mobile-annotations.md#mobile-details)
- [View Context Stamps and Details](../tutorials/mobile-annotations.md#view-context-stamps-and-details) for web views, user-generated content, and non-native content

---

## 4. Images, graphics, and other media

- [ ] **Functional images and icon-only buttons have an accessible label**
	- Decorative images should be hidden from screen readers, not given empty labels that still get focused.
- [ ] **Important text is not baked into images**

> 	- It can't be translated, scaled with Dynamic Type, or copied.

Reviewer 2: "Scaled up" instead of "Dynamic Type"

- [ ] **Advanced gestures are avoided unless essential**
	- Multi-finger taps, two-finger rotation, and path-based gestures should not be the only way to do something. If they are essential, document the [accessible alternatives](https://tetralogical.com/blog/2023/03/17/foundations-pointer-gestures/#examples-of-accessible-alternatives).
- [ ] **Drag-and-drop has a non-dragging alternative**
	- Per [SC 2.5.7 Dragging Movements](https://www.w3.org/WAI/WCAG22/Understanding/dragging-movements.html). Reorder lists with up/down buttons, move items with a menu, etc.
- [ ] **Custom components support the VoiceOver scrub gesture (iOS)**
	- On iOS, the two-finger scrub gesture (sometimes called the Z-shape gesture) acts as a back/dismiss for VoiceOver users — it can navigate backward, dismiss modals, and collapse custom components. Stock components handle this automatically. If you're building custom components that should be dismissable or collapsible, wire up scrub support manually.
- [ ] **Custom gestures are exposed as VoiceOver and TalkBack Custom Actions**

> 	- So screen reader users can discover and trigger them without performing the gesture.

Reviewer 2: Custom actions are also exposed to other accessibility services, including speech recognition, switch control, and keyboard nav. Users who need those tools also need alternatives to gestures.

	- Per [SC 1.3.4 Orientation](https://www.w3.org/WAI/WCAG22/Understanding/orientation.html), don't lock orientation unless it's essential (e.g., a piano app). Many users mount their device in a fixed orientation for accessibility reasons. _All functionality and content must remain available across orientations._
- [ ] **Layouts adapt to [iOS' Dynamic Type](https://developer.apple.com/documentation/uikit/scaling-fonts-automatically) and [Android's Font Size Scaling](https://support.google.com/accessibility/android/answer/11183305)**
	- Test at the largest accessibility text sizes per [SC 1.4.4 Resize Text](https://www.w3.org/WAI/WCAG22/Understanding/resize-text.html). Content should reflow without truncation or overlap.
- [ ] **Viewport zoom and viewport resize don't break the layout**
	- Magnifier, browser zoom inside web views, and split-screen / multitasking on tablets all change the available viewport. No content overlap, no obscured controls.
- [ ] **Reduce Motion is honored**
	- Annotate any motion-heavy transition with the Reduced motion device setting so engineering knows to gate it.
- [ ] **Voice Control labels match visible text**

> 	- Voice Control on iOS and Voice Access on Android let users speak the name of a control. The visible text and the programmatic accessibility label must match — if they're visually similar but programmatically different (or vice versa), voice users can't activate the control.

Reviewer 2: "Voice users can't activate the control  as easily." 

There are workarounds, it's just more friction for a speech user if they have to use an alternate method to activate the control. 

Also maybe reference `userInputLabels` again?

### Annotations that can help
[Device setting (User Interactions)](../tutorials/user-interactions.md#device-setting): Reduced motion, Viewport zoom, Voice input, Viewport resize, Orientation, Shake device

---

## 8. Forms

- [ ] **Inputs have a persistent visible label, not just a placeholder**
	- Per [SC 3.3.2 Labels or Instructions](https://www.w3.org/WAI/WCAG22/Understanding/labels-or-instructions.html).
- [ ] **Keyboard type matches the input**
	- Email, number, phone, URL keyboards reduce errors and effort.
- [ ] **Autofill, autocomplete, and content type hints are set**
	- iOS `textContentType`, Android `autofillHints`. Helps password managers and reduces redundant entry per [SC 3.3.7 Redundant Entry](https://www.w3.org/WAI/WCAG22/Understanding/redundant-entry.html) and [SC 1.3.5 Identify Input Purpose](https://www.w3.org/WAI/WCAG22/Understanding/identify-input-purpose.html).
- [ ] **Errors are announced, indicated with text, and tied to the field**
	- Identify what went wrong using text, not just color or an icon.

> 	- Place inline errors next to or below the field with the error. For form-level errors (e.g., a submission summary), surface them at the top or bottom of the form with line items linking to each affected field.

Reviewer 2: Take out "with line items linking to each affected field." Doing this successfully relies on focus manipulation which is brittle AF.

- [ ] **Custom controls expose proper accessibility actions**
	- For example, a custom slider should respond to increment/decrement actions, not just drag.
- [ ] **Keyboard shortcuts don't override platform or assistive tech shortcuts**
	- Check against VoiceOver, TalkBack, Switch Control, and Voice Control shortcut tables before claiming a key combo. See the [Keyboard shortcut annotation guidance](../tutorials/user-interactions.md#keyboard-shortcut).
- [ ] **No single-character shortcuts are introduced for new features**
	- Per [SC 2.1.4 Character Key Shortcuts](https://www.w3.org/WAI/WCAG22/Understanding/character-key-shortcuts.html). If you must, allow users to remap or disable them. Note: long-press gestures and platform-provided accessibility features don't meet WCAG's definition of a keyboard shortcut, so they're not in scope here.

### Annotations that can help
- [Ordering](https://gh.io/annotation-tutorial-ordering) (Focus Order, Arrow Stop)
- [Keyboard shortcut (User Interactions)](../tutorials/user-interactions.md#keyboard-shortcut)

> 

Reviewer 2: Resource to add: Workday's accessibility team's how-to on mobile keyboard nav and testing.

https://github.com/workday-accessibility/accessibility-eval/blob/main/keyboard.md

	- Use [View Context Stamps](../tutorials/mobile-annotations.md#view-context-stamps-and-details) so testers know where annotation handoff happens. For web view content specifically, run it through the [Designer](./designer-checklist.md) and [Engineering](./engineering-checklist.md) checklists.

---

## 12. Platform functions

The [Platform function annotations](../tutorials/user-interactions.md#platform-function) flag built-in behaviors that can disorient users if they're not designed carefully.

- [ ] **Loading states have clear visual and programmatic feedback**
	- Use live accessibility announcements, not just a spinner. Per [SC 4.1.3 Status Messages](https://www.w3.org/WAI/WCAG22/Understanding/status-messages.html).
- [ ] **Automatic transitions are avoided or controllable**
	- System-initiated screen, state, or content changes should be paused, slowed, or dismissible. Common offenders: onboarding carousels, auto-advancing tutorials. Per [SC 3.2.1 On Focus](https://www.w3.org/WAI/WCAG22/Understanding/on-focus.html) and [SC 3.2.2 On Input](https://www.w3.org/WAI/WCAG22/Understanding/on-input.html).
- [ ] **Opening external content is announced**

> 	- "Opens in browser" or "Opens in a new window" warnings prevent disorientation, especially when leaving an app for a web view.

Reviewer 2: "Especially when leaving an app for a ~web view~ browser." 

Web views are still part of an app.

# Mobile Checklist

This checklist summarizes accessibility considerations specific to iOS and Android, drawing from [Web Content Accessibility Guidelines (WCAG) 2.2](https://www.w3.org/TR/WCAG22/), the [W3C Guidance on Applying WCAG 2.2 to Mobile Applications](https://www.w3.org/TR/wcag2mobile-22/), and best practices identified by GitHub.

Native mobile has its own set of accessibility patterns, APIs, and user expectations that don't map cleanly to the web. Use this checklist when designing or building **native mobile experiences** on iOS and Android. You don't need to also run through the [Designer Checklist](./designer-checklist.md) or [Engineering Checklist](./engineering-checklist.md) for a native app. 

> Across both platforms, the underlying fuser needs are the same: people must be able to perceive, operate, and understand the experience equivalently, even when the patterns and conventions differ. Differences in *how* something is implemented across iOS and Android are expected and fine. Differences in *whether* a user can access information, complete a task, or form an understanding are not.

Reviewer 2: 1. Typo: fuser
2. My intent was to draw the line between web accessibility and mobile and how they're fundamentally different but seek to meet the same goals.

Author: That makes more sense 😅 how's this?

> Web and native mobile look different, but they're chasing the same goal: people must be able to perceive, operate, and understand the experience regardless of platform. The patterns, APIs, and conventions diverge, but the underlying user needs don't. This checklist focuses on how that goal gets met on native — where stock controls, system settings, and gesture conventions do a lot of the work that markup and ARIA do on the web.

Reviewer 2: Yes! Love it!
